execute/exectypes/type_optimizer.go

```diff
@@ -0,0 +1,209 @@
+package exectypes
```

Comment:
could you move this out of the types package?

Reply:
Maybe a new internal package, since nothing here really needs to be exported beyond the execute.